rp2: Add support for named pins and alternate functions.

iabdalkader opened the pull request:
This PR adds named pins and alternate functions support in the `rp2` port, something that's been missing for a very long time, and is more important now especially with recent changes that added support for CYW43-controlled pins, which made pin management (in `machine_pin.c`) much more complex.

* The first commit refactors `machine_pin.c` to handle externally controlled GPIO pins more generically, by removing all CYW43-specific code from `machine_pin.c` to `machine_pin_cyw43.c`, and adding hooks to initialise, configure, read and write external pins, that can be implemented by any driver that controls external pins. External pin control is enabled by defining the number of external GPIOs in `MICROPY_HW_PIN_EXT_COUNT`, otherwise only CPU pins are used.
* The following commit adds support for generating named pin mappings for all pins including the CPU pins, board-defined pins (in `pins.csv`), LED pin and externally controlled pins (again enabled by `MICROPY_HW_PIN_EXT_COUNT`). CPU pins are mapped to `pin_GPIO<n>`, externally-controlled pins are mapped to `pin_WL_GPIO<n>`, and defined conditionally (up to 10 pins, and can be expanded in the future), and they are non-const to allow `machine_pin_xxx.c` to write the pin fields. Both CPU and external pins are generated even if there's no board CSV file, if one exists it will just be added to board pins. Note the prefix for external pins `WL_` is fixed for now (very hard to generate that with macros, probably should use `EXT_` and use custom naming in `pins.csv`). 
* The last 2 commits, implement external pin control for Nina module as en example (in addition to the CYW43 controls, which I left in the `rp2` port, for now, but can be moved to the cyw43 driver).
* There's special handling for the LED pin (if one is defined), by default "LED" is mapped to a CPU pin (default pico pin), and on CYW43 builds, "LED" is mapped to a CYW43 pin (if defined). However, the LED, and any externally controlled pin, can also be mapped in `pins.csv` same as CPU pins, for example: `LEDR,WL_GPIO0`. It's possible to remove this special case, if all boards add `LED,<pin>` to their `pins.csv`.

For testing, I used the following simple script:

```python
from machine import Pin
def print_pin(*args, **kwargs):
    try:
        print(Pin(*args, **kwargs))
    except Exception as e:
        print (e)

print_pin("LED", Pin.OUT)                       # On Pico-W this WL_GPIO0, on Pico this GPIO6
print_pin("RESET", Pin.OUT)                     # board pin defined for Nano
print_pin("WL_GPIO2", Pin.OUT)                  # Works on PICO-W
print_pin("GPIO9", Pin.ALT, alt=Pin.AF_PWM)     # PWM AF
print_pin("GPIO10", Pin.ALT, alt=Pin.AF_CLOCK)  # Should fail, no CLOCK AF on GPIO10
````

Running this on Nano-RP2040 board (non-cyw43 build), with a custom board pin (`RESET`) and LED pin on `GPIO6`, I get the following output:
```bash
Pin(GPIO6, mode=OUT, ext_pin=false)
Pin(GPIO3, mode=OUT, ext_pin=false)
Unknown named pin "WL_GPIO2".
Pin(GPIO9, mode=ALT, alt=PWM, ext_pin=false)
Invalid pin af: 8.
```

Running on the Pico-W ( no board reset pin, LED controlled by `WL_GPIO0`):

```bash
Pin(WL_GPIO0, mode=OUT, ext_pin=true)
Unknown named pin "RESET".
Pin(WL_GPIO2, mode=OUT, ext_pin=true)
Pin(GPIO9, mode=ALT, alt=PWM, ext_pin=false)
Invalid pin af: 8.
```
Note: even though AF `unit` is not really needed/used, I kept them around for future boards (re-adding them will be a lot of work).

ubidefeo commented:
👍🏼 
beautiful

iabdalkader commented:
Example generated `pins_PICO_W.c`:
```c
const machine_pin_af_obj_t pin_GPIO0_af[] = {
  AF(1, SPI , 0), // SPI0_RX
  AF(2, UART, 0), // UART0_TX
  AF(3, I2C , 0), // I2C0_SDA
  AF(4, PWM , 0), // PWM0_A
  AF(5, SIO , 0), // SIO
  AF(6, PIO0, 0), // PIO0
  AF(7, PIO1, 1), // PIO1
  AF(9, USB , 0), // USB_OVCUR_DET
};

const machine_pin_obj_t pin_GPIO0 = PIN(0, GPIO0, 0, 8, pin_GPIO0_af);
...

#if (MICROPY_HW_PIN_CYW43_COUNT > 0)
// const machine_pin_af_obj_t pin_WL_GPIO0_af[] = {
// };

machine_pin_obj_t pin_WL_GPIO0 = PIN(0, WL_GPIO0, 1, 0, NULL);
#endif
...

STATIC const mp_rom_map_elem_t pin_cpu_pins_locals_dict_table[] = {
  { MP_ROM_QSTR(MP_QSTR_GPIO0), MP_ROM_PTR(&pin_GPIO0) },
  ...
  { MP_ROM_QSTR(MP_QSTR_GPIO29), MP_ROM_PTR(&pin_GPIO29) },
};
MP_DEFINE_CONST_DICT(pin_cpu_pins_locals_dict, pin_cpu_pins_locals_dict_table);


STATIC const mp_rom_map_elem_t pin_board_pins_locals_dict_table[] = {
  #if (MICROPY_HW_PIN_CYW43_COUNT > 0)
  { MP_ROM_QSTR(MP_QSTR_WL_GPIO0), MP_ROM_PTR(&pin_WL_GPIO0) },
  #endif
  ...
  #if (MICROPY_HW_PIN_CYW43_COUNT > 9)
  { MP_ROM_QSTR(MP_QSTR_WL_GPIO9), MP_ROM_PTR(&pin_WL_GPIO9) },
  #endif
  #ifdef NAMED_PIN_LED_PIN
  { MP_ROM_QSTR(MP_QSTR_LED), MP_ROM_PTR(&NAMED_PIN_LED_PIN) },
  #endif
};
MP_DEFINE_CONST_DICT(pin_board_pins_locals_dict, pin_board_pins_locals_dict_table);
```

rkompass commented:
Very nice.
Now assuming this is merged, shouldn't
`Pin(GPIO9, mode=ALT, alt=4, cyw43_pin=false)`
instead report
`Pin(GPIO9, mode=ALT, alt=AF_PWM, cyw43_pin=false)`
?

iabdalkader commented:
> instead report
> `Pin(GPIO9, mode=ALT, alt=AF_PWM, cyw43_pin=false)`

Sure, it now does.

iabdalkader commented:
@dpgeorge I want to extend this into something more generic, to support Nina GPIOs too, which work exactly the same as the CYW43 GPIOs (via ioctls). For example externally controlled pins can be enabled by `MICROPY_HW_PIN_COPROC/EXT/WL_COUNT`, use a configurable prefix (ex: `WL_GPIO`), the functions that actually call the ioctl can be implemented in another file, for example `machine_pin_cyw43.c` (which should also reduce `machine_pin.c` significantly), rename `is_cyw43` to `is_coproc/ext`. This should allow Nano `pins.csv` to define the RGB LED, for example `LEDR, WL_GPIO0`. Would that be a good idea ?

iabdalkader committed:
rp2/machine_pin: Refactor the machine.Pin class for external GPIO.

Handle externally controlled GPIO pins more generically, by removing all
CYW43-specific code from `machine_pin.c`, and adding hooks to initialise,
configure, read and write external pins.  This allows any driver for an
on-board module which controls GPIO pins (such as CYW43 or NINA), to
provide its own implementation of those hooks and work seamlessly with
`machine_pin.c`.

rkompass commented:
@iabdalkader @dpgeorge just a question, perhaps not really relevant here:
Is there a plan to add the drive option to rp2 pins? And perhaps an inversion option, as both may be set by registers.
I recently made the following to try out the relaxation oscillator mentioned in the manual using the clock alternative function in connection with pin inversion (and hardwired RC-delay):

```
# Improved micropython Pin class for RPi2040
#
# (c) 2022 by rkompass
#
# MIT License
#
# Adds keyword arguments inv_io, drive and slewfast to init() and initializer.
# Adds trueval() method to get true (not possibly inverted) pin value from the pin pad.
#
#    inv_io = (1, 0) invert input
#           = (2, 0) input permanently low
#           = (3, 0) input permanently high
#           = (1, 1) invert input and output
#          ...
#           = (3, 3) input and output permanently high
#
#    drive = 0  ->  2mA  output drive strength
#          = 1  ->  4mA
#          = 2  ->  8mA
#          = 3  -> 12mA
#
#    slewfast = 1        output amplifier slew rate set to fast
#
#
# Example:
# --------
#
# from rp2_nupin import nuPin
# 
# led = nuPin(25, Pin.OUT, value=0, inv_io=(1,1)) # set pin low on creation but invert in- and output
#                                                 # led goes on
# print('led.value():', led.value())              #  --> 0
# print('led.trueval():', led.trueval())          #  --> 1
# led.init(inv_io=(0,0))                          # led goes off
# print('led.value():', led.value())              #  --> 0
# print('led.trueval():', led.trueval())          #  --> 0


from machine import Pin, mem32
from array import array

class nuPin(Pin):

    def __init__(self, *args, **kwargs):
        self.pin = args[0]
        super().__init__(*args)
        newkargs = {'inv_io':None, 'drive':None, 'slewfast':None}
        for k in newkargs:
            if k in kwargs:
                newkargs[k] = kwargs[k]
                del kwargs[k]
        super().init(**kwargs)
        self.init(**newkargs)

    def init(self, mode=None, pull=None,*, value=None, alt=None, inv_io=None, drive=None, slewfast=None):
        if alt is not None:
            super().init(mode, pull, value=value, alt=alt)  # do super().init() with mode, pull, value and alt
        else:
            super().init(mode, pull, value=value) # Pin.init() does not tolerate an alt=None argument, so avoid it
        if inv_io is not None: # adjust pin ctrl register if necessary
            if not isinstance(inv_io, (tuple, list)) or len(inv_io) != 2:
                raise ValueError('inv_io argument expects 2 elements for input / output inversion')
            ctrlr = mem32[0x40014004 + 8*self.pin]
            ctrlr &= 0xfffcfcff
            ctrlr |= (inv_io[0] & 3) << 16 | (inv_io[1] & 3) << 8 # 1..invert, 2..permanent low, 3..permanent high
            mem32[0x40014004 + 8*self.pin] = ctrlr       
        padr = mem32[0x4001c004 + 4*self.pin]   # adjust pin drive and/or slew rate in Pin pad register if necessary
        if drive is not None:
            padr &= 0xffffffcf
            padr |= (drive & 3) << 4    # drive may be 0 (2mA), 1 (4mA), 2 (8mA) or 3 (12 mA)
        if slewfast is not None:
            padr &= 0xfffffffe
            padr |= (slewfast & 1)
        mem32[0x4001c004 + 4*self.pin] = padr

    @micropython.viper
    def trueval(self) -> int:              # returns the input signal from pin pad before gpio overrides
        pin = int(self.pin)
        p = ptr32(0x40014000)
        return p[2*pin] >> 17 & 1
        
    def __repr__(self):
        str0 = repr(super()).split('>')[1][2:-1]
        ctrlr = mem32[0x40014004 + 8*self.pin]
        padr = mem32[0x4001c004 + 4*self.pin]
        inv_io = (ctrlr >> 16 & 3, ctrlr >>  8 & 3)
        drive = padr >> 4 & 3
        slewfast = padr & 1
        r_inv  = ', inv_io=({:d},{:d})'.format(inv_io[0], inv_io[1]) if inv_io != (0, 0) else ''
        r_driv = ', drive={:d}'.format(drive) if drive != 1 else ''
        r_slew = ', slewfast=True' if slewfast else ''
        return 'nu{:s}{:s}{:s}{:s})'.format(str0, r_inv, r_driv, r_slew)
    
# from time import sleep
# led = nuPin(25, Pin.OUT, value=0, inv_io=(1,1), drive=2, slewfast=False) # set pin low on creation but invert in- and output
# print('led.value():', led.value())
# print('led.trueval():', led.trueval())
# print(repr(led))
# sleep(2)
# led.init(value=1, inv_io=(0,0), drive=0, slewfast=True)
# print('led.value():', led.value())
# print('led.trueval():', led.trueval())
# print(repr(led))
# # sleep(2)
# # led.init(value=0, drive=1, slewfast=False)
```
I'm fine with this solution, but other ports **have** the drive option.
The inversion option would be nice to have built in, but perhaps confusing. Therefore I added the .trueval() function.
The slewfast option did not show any difference for me even with 250MHz clock output so I suppose it is not useful.

iabdalkader committed:
rp2/machine_pin: Add support for named pins and alternate functions.

This commit adds support for generating named pin mappings for all pins
including CPU, board-defined, LED and externally controlled pins.  CPU pins
are mapped to `pin_GPIO<n>`, externally-controlled pins are mapped to
`pin_EXT_GPIO<n>`, and defined conditionally (up to 10 pins, and can be
expanded in the future), and they are non-const to allow `machine-pin.c` to
write the pin object fields.  Both CPU and externally controlled pins are
generated even if there's no board CSV file; if one exists it will just be
added to board pins.

iabdalkader committed:
drivers/ninaw10: Implement machine.Pin external pin controls.

iabdalkader committed:
rp2/boards: Add pin CSV files to board definitions.

Pin defines are:
- For Pico define board pins and the default LED pin (WL_GPIO25).
- For Pico-W define board pins, external pins and the default
  LED pin (WL_GPIO0).
- For the Nano-RP2040, define board pins, external pins and
  the default LED pin (GPIO25)
- For all other boards, the pins.csv defines the LED pin (if any)
  for backwards compatibility with code that assumes there's always
  an LED pin.

iabdalkader commented:
>  Is there a plan to add the drive option to rp2 pins? And perhaps an inversion option, as both may be set by registers.

Not really, I'm only interested in cleaning up `machine_pinc.c` and adding support for named pins. However, I reworked this PR to refactor `machine_pin.c` (see updated first comment), the external pin control functions are located in another file now (`machine_pin_cyw43.c`), if merged it might be easier and more acceptable to add more logic there.

dpgeorge commented:
Overall this looks pretty good, but I didn't do a detailed review yet.

One thing I'd say is that the "AF" naming is a legacy thing from old stm32 days, where it means "alternate function".  It's somewhat been replaced by "ALT" and I think we should just use "ALT" everywhere for such naming and not use "AF".  So the constants would be "Pin.ALT_PWM" etc.

> I want to extend this into something more generic, to support Nina GPIOs too, which work exactly the same as the CYW43 GPIOs (via ioctls)

That makes sense.  But for now let's just concentrate on this PR as-is to get it over the line.

> Also while we're at it, would it be better to merge `pin.h` and (the functions in) `pin.c` with `machine_pin.h` and `machine_pin.c` ?

`pin.[ch]` should be a low-level C interface to the underlying GPIO functionality.  Then `machine_pin.[ch]` are the Python bindings.  At least that's the general rule.  (And ideally bindings would go in `extmod/machine_pin.[ch]` and be the same across all ports, but for pin that's quite hard to do so don't do that now.)

dpgeorge commented:
> Is there a plan to add the drive option to rp2 pins? And perhaps an inversion option, as both may be set by registers.

It does make sense to eventually add `drive` support, since that's available on other ports.  We could also add inversion.  But both of those things should be done in separate PRs.

iabdalkader commented:
> One thing I'd say is that the "AF" naming is a legacy

I can rename that, and fix the scripts but would rather wait for a full review to fix anything else while I'm at it.

> should be a low-level C interface to the underlying GPIO functionality. Then machine_pin.[ch] are the Python bindings.

I actually meant where do `*_find_named_pin/af/af_by_index` etc.. should go ? For now I just put them in `machine_pin.c`, because they're not low-level, and to avoid extra files `pin.h` and `pin_named.c` for 3-4 functions, but can move to separate files.

> That makes sense. But for now let's just concentrate on this PR as-is to get it over the line.

I don't have any more changes, to recap I basically:

1) Refactored `machine_pin.c` to make external pins more generic and because I don't think cyw43-specific code belongs in there regardless of any other changes, so moved all CYW43 code to `machine_pin_cyw43`, renamed things like `is_cyw43` to `is_ext`. This looks like a good decision now since you're okay with adding more `cyw43` GPIO options, so now they won't clutter `machine_pin.c`

2) Add the alternate pin function generation.

3) Added a similar file in `drivers/ninaw10/machine_pin_nina.c`. This can be dropped but it helps to see the bigger picture with another board.

jimmo reviewed (commented):
>  It's possible to remove this special case, if all boards add LED,<pin> to their pins.csv.

I think we should do this -- it looks like it will remove a bunch of special handling in make-pins.py and overall make things simpler and more obvious -- the whole point of the pins.csv mechanism is to give board-specific names to pins.

> very hard to generate that with macros, probably should use EXT_ and use custom naming in pins.csv

Yes, similarly, I'd much rather move the "magic" to pins.csv. We may want to support other types of GPIO expanders at this level, and "ext" matches the naming used for e.g. machine_pin_ext_set.

---

Thanks @iabdalkader this is a great improvement!

FWIW (and I don't think we should address in this PR) but at some point we should change rp2's `mp_hal_pin_obj_t` to be the actual pin object rather than int (like it is on stm32).

jimmo commented:
> Yes, similarly, I'd much rather move the "magic" to pins.csv. We may want to support other types of GPIO expanders at this level, and "ext" matches the naming used for e.g. machine_pin_ext_set.

More detail...

Background:
- In exactly the same way as the stm32 port, we now have auto-generated "cpu" pins and board-configured "board" pins which are aliases to cpu pins (defined in pins.csv).
- Novel to the rp2 port (but hopefully could be added to other ports), we now have "external" pins which this PR is making "board" pins.
- For backwards compatibility, specifically for the Pico W, the external pins need to be constructable as `machine.Pin("WL_GPIOn")`

I think I would argue that these external pins are not board pins -- board pins are typically what the silkscreen would label them as or what they do functionally. i.e. "LED" is a good example. So by that logic they're "cpu" pins.

We could add a new category of pins -- `machine.Pin.ext` ? -- but I think it's simpler in this case to consider the external pins as "cpu", and call them "EXT_GPIOn". Then specifically for the Pico W, create "board" pin aliases for these as backwards compatibility, e.g. "WL_GPIOn" (as well as "LED").

So pico_w/pins.csv would be

```
WL_GPIO0,EXT_GPIO0
WL_GPIO1,EXT_GPIO1
...
LED,EXT_GPIO0
```

iabdalkader commented on ports/rp2/machine_pin.h:
```diff
@@ -0,0 +1,94 @@
+    MACHINE_PIN_MODE_ALT = 3
+};
+
+enum {
```
Sure.

iabdalkader commented:
I made the following updates based on your suggestions @jimmo @dpgeorge 

* Renamed constants to `ALT` (for example `Pin.ALT_PWM` etc...).
* Use `gpio_function` enum provided in pico-sdk.
* Removed special cases for LED pin (now it's just defined in `pins.csv`).
* Add `pins.csv` to every board with at least the default Pico LED defined (this is required now for backwards compatibility).
* Add a `pins.csv` for Pico-W with the pins jimmo suggested.

* Consider the external pins as "cpu", and call them "EXT_GPIOn". 

@jimmo  technically, external pins are board-specific (not CPU), they could be driven by a WiFi module, or any other chip on-board, although not sure if the distinction really matters, because it's all handled internally, to users they are just pins, correct ? But we can revisit this later and make a new group if needed.

I'm also attaching the generated pins for Pico-W.
[pins_PICO_W.c.txt](https://github.com/micropython/micropython/files/9877263/pins_PICO_W.c.txt)

jimmo commented on ports/rp2/boards/PICO_W/pins.csv:
```diff
@@ -0,0 +1,13 @@
+WL_GPIO0,EXT_GPIO0
```
The Pico W only has 3 external pins.

```c
#define CYW43_WL_GPIO_COUNT 3
```

iabdalkader replied:
Right :man_facepalming:  sorry

iabdalkader replied:
FWIW this is `board` table with only 3 pins:
```c
STATIC const mp_rom_map_elem_t pin_board_pins_locals_dict_table[] = {
  #if (MICROPY_HW_PIN_EXT_COUNT > 0)
  { MP_ROM_QSTR(MP_QSTR_WL_GPIO0), MP_ROM_PTR(&pin_EXT_GPIO0) },
  #endif
  #if (MICROPY_HW_PIN_EXT_COUNT > 1)
  { MP_ROM_QSTR(MP_QSTR_WL_GPIO1), MP_ROM_PTR(&pin_EXT_GPIO1) },
  #endif
  #if (MICROPY_HW_PIN_EXT_COUNT > 2)
  { MP_ROM_QSTR(MP_QSTR_WL_GPIO2), MP_ROM_PTR(&pin_EXT_GPIO2) },
  #endif
  #if (MICROPY_HW_PIN_EXT_COUNT > 0)
  { MP_ROM_QSTR(MP_QSTR_LED), MP_ROM_PTR(&pin_EXT_GPIO0) },
  #endif
};
MP_DEFINE_CONST_DICT(pin_board_pins_locals_dict, pin_board_pins_locals_dict_table);
```

jimmo reviewed (commented):
Thanks for updating! Looks good!

> @jimmo technically, external pins are board-specific (not CPU), they could be driven by a WiFi module, or any other chip on-board, although not sure if the distinction really matters, because it's all handled internally, to users they are just pins, correct ? But we can revisit this later and make a new group if needed.

Yeah absolutely. I'm just saying they're more "cpu" than "board" (simply by the rationale that "board" means "defined in board.csv").

jimmo commented:
Sorry to throw a bit of a spanner in the works... I was looking into how this works on STM32 -- A pin will only be in "cpu" if it exists in pins.csv... i.e. we only make a "cpu" entry if the pin can be accessed on this board. In most cases board pins have all different names, but in some cases there are boards that label the CPU name on the silkscreen (e.g. I'm looking at a Nucleo right now that has the Arduino header plus extra pins labeled "PB13", etc). (In that case we add e.g. `PB13,PB13` to pins.csv).
The overall idea is that "cpu" only contains pins that are accessible on the board, and "board" contains their true names. (In other words: my understanding is that you cannot construct a pin instance for a pin that isn't in pins.csv).

So for rp2040...
a) Should we include non-accessible pins in "cpu" ? e.g. the [Tiny2040](https://shop.pimoroni.com/products/tiny-2040?variant=39560012234835) only has 12 pins broken out.
b) The Pico (and Pico W) labels the pins as "GPn", so I would expect therefore to write `machine.Pin.board.GPn` or `machine.Pin("GPn")`.
c) Is there any reason the CPU names have to be "GPIO_N"? (Could they be shorter and save QSTR space?). We don't have any backwards compatibility to maintain here. I think the only things to maintain are "WL_GPIOn" (only Pico W) and "LED" (all boards).
d) Other boards have all sorts of other labels (e.g. the Tiny2040 calls them ("0", "1", ..., "A1", ...).
e) The nano connect labels the pins as "An", "Dn", "TX1", "RX0", "REC", so this PR should probably include those in the pins.csv for this board.

My thoughts:
a) Really the only benefit here is flash saving, but doing this would make it consistent with STM32 (and my longer term goal is to unify this whole process across ports).
b+c) You should definitely be able to use the silkscreen name. So as a simplification we could make "GPn" the CPU name and then Pico/PicoW would be sorted. Otherwise we should add GPn to pins.csv for both boards.
d) It's tricky when the board/silkscreen names don't start with a letter! I think on this board having cpu.GPn and board.Pn/board.An would work?
e) Same as Pico, the board names should match the silkscreen.

I don't want to create a bunch of extra work for you to create pins.csv for each existing board (although it needs to be done eventually). So perhaps this PR should just focus on Pico/PicoW/Nano, and we include all CPU pins by default (and address that later once all boards have fully defined pins.csv).

iabdalkader commented:
> In other words: my understanding is that you cannot construct a pin instance for a pin that isn't in pins.csv).
> a) Should we include non-accessible pins in "cpu" ?

I generate them all because `machine_pin.c` actually lets you index into the `cpu` pins  (the pin array not the named pin `dict`) with an `int`, so they must all exist and be defined in order, but if we change then it shouldn't be possible/allowed to access pins with an index, correct ? Note this is also a breaking change (not being able to get a pin with `Pin(0)` to be clear).

> b+c) You should definitely be able to use the silkscreen name. So as a simplification we could make "GPn" the CPU name and then Pico/PicoW would be sorted. Otherwise we should add GPn to pins.csv for both boards.

I'd choose option 2 (adding `GPn` to `pins.csv`) because a) That's the actual CPU/chip pin name `GPIO<n>` as it is in the datasheet (so just like `stm32`'s `P<c><n>` for example) b) silkscreen is board-specific, and space restricted, if they make a new Pico they may call it `GPIO` again b) It matches `WL_GPIO<n>`

> d) It's tricky when the board/silkscreen names don't start with a letter! I think on this board having cpu.GPn and board.Pn/board.An would work?

You'd have to append something I think so yes, `P<n>` maybe in the `pins.csv`.

> but doing this would make it consistent with STM32 (and my longer term goal is to unify this whole process across ports).

I'm all for consistency between ports, but in this case the STM32 port is very old, just wanted to point that out, might want to survey all the other ports, see what's really common right now.

iabdalkader commented:
> Just to be clear -- this PR still allows Pin(int) -- you're saying that if we removed "unused" pins, then it would require breaking that (because the array would now be sparse)?

Exactly.

> OK so cpu.GPIOn and board.GPn then? (And keep board.WL_GPIOn for Pico W).

Yes again on spot. Although, side note that I have to mention, I actually forgot to add those 2 `dict`s to the Pin object, but will do that now in my next round of fixes.

> STM32 is the reference. SAMD, mimxrt, renesas all started with STM32's version. nrf, teensy, zephyr all also support this, but with differences.

I usually look there first too, some practices there have evolved in better? ways I think, for example `AF` -> `ALT` pointed out by dpgeorge.

iabdalkader commented:
@jimmo Updated, all pins are defined for Pico, Pico-W and Nano-RP2040, for all other boards only the LED pin is defined (some didn't have any, so removed their `pins.csv` file). `machine.Pin.board` and `machine.Pin.cpu` now have the discussed pins.

jimmo commented:
Thanks @iabdalkader - looks good to me, thanks for making the changes, this is a great improvement for the rp2 port.

iabdalkader commented:
Thanks for the review!

I have a few more thoughts about, and they don't necessarily have to be addressed in this PR, just wanted to share for future updates:

* I think the cyw43-specific state in `machine_pin_obj_t` (`is_output` and `last_output_value`, etc...) should be moved to `machine_pin_cyw43.c`, and maintained there internally, especially if more state is added for `cyw43` pins (drive, pulls etc..). This will also allow external pins to be declared `const`.
* External pins can be defined by just defining them in `pins.csv`  (not using `MICROPY_HW_PIN_EXT_COUNT`) this will allow any number of external pins, and only the ones defined will show up (unconditionally) in the generated tables.
* If the above is implemented, then pin objects will always have the `is_ext` flag (which will always be 0 for the `const` non-external pins). This may add small overhead though, because `is_ext_pin(x)` won't be optimized out, if there are no external pins.

dpgeorge commented:
> OK so cpu.GPIOn and board.GPn then?

Yes I agree with this: the RP2040 datasheet calls them GPIOn, so that's what their cpu name is.  And the silkscreen has them as GPn, so that's the board name.

dpgeorge commented on ports/rp2/machine_pin.c:
```diff
@@ -240,12 +196,18 @@ STATIC void machine_pin_print(const mp_print_t *print, mp_obj_t self_in, mp_prin
+                mp_printf(print, ", alt=%q", af->name);
+            }
         }
+        mp_printf(print, ", ext_pin=%b", is_ext_pin(self));
```
I think this line should be wrapped in `#if MICROPY_HW_PIN_EXT_COUNT` / `#endif`, so that on boards that do not have any ext pins (eg PICO) the `ext_pin` part is never printed.

Also, is `ext_pin` even needed to be printed?  This is not a keyword argument accepted by the `Pin()` constructor. The name of the pin (starts with `EXT_`) is enough to indicate it's external.  Furthermore, `%b` uses lowercase true/false, whereas it should print uppercase to match the Python True/False objects.

iabdalkader replied:
Removed.

dpgeorge commented on ports/rp2/machine_pin.c:
```diff
@@ -594,47 +490,59 @@ STATIC const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_ALT_SIO), MP_ROM_INT(GPIO_FUNC_SIO) },
+    { MP_ROM_QSTR(MP_QSTR_ALT_PIO0), MP_ROM_INT(GPIO_FUNC_PIO0) },
+    { MP_ROM_QSTR(MP_QSTR_ALT_PIO1), MP_ROM_INT(GPIO_FUNC_PIO1) },
+    { MP_ROM_QSTR(MP_QSTR_ALT_CLOCK), MP_ROM_INT(GPIO_FUNC_GPCK) },
```
I suggest this should be `MP_QSTR_ALT_GPCK` to match the pico-sdk name, like all other constant in this group (is it GPCK in the datasheet?).

iabdalkader replied:
Changed to `GPCK`. The datasheet name is `CLOCK GP(IN/OUT)n`, depending on the pin, and there's only one alternate function for in or out.

dpgeorge commented on ports/rp2/machine_pin.c:
```diff
@@ -261,155 +223,101 @@ static const mp_arg_t allowed_args[] = {
+            // Configure alternate function.
+            mp_uint_t af = args[ARG_alt].u_int;
+            if (machine_pin_find_af(self, af) == NULL) {
+                mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("Invalid pin af: %d."), af);
```
Please rename all uses of `af` to `alt` (eg `machine_pin_find_alt()`).

Also error messages should start with a lowercase letter (unless it's a class name).

iabdalkader replied:
Renamed and fixed error messages.

dpgeorge commented on ports/rp2/machine_pin.c:
```diff
@@ -261,155 +223,101 @@ static const mp_arg_t allowed_args[] = {
-        #endif
         if (!self) {
-            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("Unknown named pin \"%s\""), name);
+            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("Unknown named pin \"%s\"."), name);
```
lowercase, no full-stop

iabdalkader replied:
FYI existing error messages start with uppercase: https://github.com/micropython/micropython/blob/master/ports/rp2/machine_pin.c#L273

Not sure if you want me to fix them all or just the ones I added.

iabdalkader replied:
Fixed all exception messages.

dpgeorge commented:
Thanks for updating!

kilograham commented:
Are we sure that "LED" actually still works on pico-w, i heard that it doesn't

dpgeorge commented:
It does.  But you can't create a PWM object out of the LED pin.

kilograham commented:
ah, apologies... i got the wrong end of the stick - i was gonna try it tomorrow, honest!

dpgeorge commented:
no worries!
